hs_core/hydroshare/utils.py

```diff
@@ -844,105 +838,65 @@ def resource_file_add_process(resource, files, user, extract_metadata=False,
-                                                  f, UploadedFile) else f)
-        else:
-            ret = ResourceFile.objects.create(content_object=resource,
-                                              file_folder=folder,
-                                              resource_file=File(f) if not isinstance(
-                                                  f, UploadedFile) else f,
-                                              fed_resource_file=None)
-
-    elif fed_res_file_name_or_path and (fed_copy_or_move == 'copy' or fed_copy_or_move == 'move'):
-        size = get_fed_zone_file_size(fed_res_file_name_or_path)
-        file_format_type = get_file_mime_type(fed_res_file_name_or_path)
-        ret = ResourceFile.objects.create(content_object=resource, resource_file=None,
-                                          fed_resource_file=None,
-                                          fed_resource_file_name_or_path=fed_res_file_name_or_path,
-                                          fed_resource_file_size=size)
-        try:
-            from_fname = fed_res_file_name_or_path
-            filename = from_fname.rsplit('/')[-1]
-            istorage = resource.get_irods_storage()
-            if resource.resource_federation_path:
-                if folder:
-                    to_fname = os.path.join(resource.resource_federation_path, resource.short_id,
-                                            'data', 'contents', folder, filename)
-                else:
-                    to_fname = os.path.join(resource.resource_federation_path, resource.short_id,
-                                            'data', 'contents', filename)
-            else:
-                if folder:
-                    to_fname = os.path.join(resource.short_id, 'data', 'contents',
-                                            folder, filename)
-                else:
-                    to_fname = os.path.join(resource.short_id, 'data', 'contents',
-                                            filename)
 
-            if fed_copy_or_move == 'copy':
-                istorage.copyFiles(from_fname, to_fname)
-            else:
-                istorage.moveFile(from_fname, to_fname)
-            # update file path now that file has been copied or moved to HydroShare
-            # proxy account space
-            ret.fed_resource_file_name_or_path = 'data/contents/{file_name}'.format(
-                file_name=filename)
-            ret.save()
+    elif source_name:
+        try:
+            # create from existing iRODS file
+            ret = ResourceFile.create(resource, None, folder=folder, source=source_name, move=move)
         except SessionException as ex:
-            # delete the file added if there is any exception
-            ret.delete()
             # raise the exception for the calling function to inform the error on the page interface
             raise SessionException(ex.exitcode, ex.stdout, ex.stderr)
```

Comment:
@alvacouch I think it is safer to include that ```ret.delete()``` if there is an exception since the exception is caused by iRODS at which point, a record may have already been created in Django database since Django DB operation and iRODS file operation are not atomic, so we don't want to leave any "gabage" in django database if there is an iRODS exception.